enhancements/kube-apiserver/stability-no-nodes.md

```diff
@@ -0,0 +1,199 @@
+---
+title: stability-no-nodes
+authors:
+  - "@deads2k"
+reviewers:
+approvers:
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - "/enhancements/this-other-neat-thing.md"  
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# Stability: No Target Nodes
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.  For instance, 
+ > 1. This requires exposing previously private resources which contain sensitive
+  information.  Can we do this? 
+
+## Summary
+
+When a cluster has no workers or the wrong number of masters, the cluster will fail to install because components like ingress cannot run.
+When this happens we need a very good error message.
+This is worthy of more than an alert because it happens without workers and has been a practical problem on many clusters,
+particularly UPI and baremetal.
+
+## Motivation
+
+Answering the same question after digging through a cluster is a waste of time.
+This came up a lot in 4.4 as baremetal clusters failed to start.
+
+### Goals
+
+1. Fail with a clear error message when a component cannot successfully create an operand because no target nodes are available.
+
+### Non-Goals
+
+1. Try to figure out why target nodes are unavailable.
+
+## Proposal
+
+Operators can know when they have zero pods and can perform basic checks to see why, including
+ 1. CNI plugin doesn't exist
+ 2. No suitable nodes found
+ 3. If I install on workers, are there any workers.
+ 
+Metrics and alerts can help, but having a clear, unsuppressable indication on an operator gives feedback in the spot
+we've trained things like the installer, CEE, and CI to look.
```

Comment:
Yes, each operator could report "I am Degraded because there are no worker nodes" but it would be simpler to just have some operator be responsible for checking whether worker nodes exist, and reporting a Degraded status when they don't.

Likewise, CNO could report "I am Degraded because there are 6 nodes and I have only deployed the network plugin to 5 of them", but if it does that, then people are just going to assign the bug to Networking, when the actual bug is "one of the nodes crashed after rebooting" or something like that. There ought to be something that independently keeps track of whether there are non-`Ready` nodes, and reports when there are, even if it can't figure out why they're not `Ready`.

Likewise, cluster-openshift-apiserver-operator ought to be able to report a status of "apiserver-reachability is Degraded".

The problem with `oc get clusteroperator` is that it answers the question "whose fault is it that the cluster is broken?", so it has a hard time being useful when the cluster is broken but nobody knows why.